src/rust/engine/options/src/parse.rs

```diff
@@ -274,49 +278,188 @@ fn format_parse_error(
+impl Parseable for bool {
+    fn parse(value: &str) -> Result<bool, ParseError> {
+        option_value_parser::bool(value).map_err(|e| format_parse_error("bool", value, e))
+    }
+
+    fn parse_list(value: &str) -> Result<Vec<ListEdit<bool>>, ParseError> {
+        option_value_parser::bool_list_edits(value)
+            .map_err(|e| format_parse_error("bool list", value, e))
+    }
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_float(value: &str) -> Result<f64, ParseError> {
-    option_value_parser::float(value).map_err(|e| format_parse_error("float", value, e))
+impl Parseable for i64 {
+    fn parse(value: &str) -> Result<i64, ParseError> {
+        option_value_parser::int(value).map_err(|e| format_parse_error("int", value, e))
+    }
+
+    fn parse_list(value: &str) -> Result<Vec<ListEdit<i64>>, ParseError> {
+        option_value_parser::int_list_edits(value)
+            .map_err(|e| format_parse_error("int list", value, e))
+    }
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_quoted_string(value: &str) -> Result<String, ParseError> {
-    option_value_parser::quoted_string(value).map_err(|e| format_parse_error("string", value, e))
+impl Parseable for f64 {
+    fn parse(value: &str) -> Result<f64, ParseError> {
+        option_value_parser::float(value).map_err(|e| format_parse_error("float", value, e))
+    }
+
+    fn parse_list(value: &str) -> Result<Vec<ListEdit<f64>>, ParseError> {
+        option_value_parser::float_list_edits(value)
+            .map_err(|e| format_parse_error("float list", value, e))
+    }
+}
+
+impl Parseable for String {
+    fn parse(value: &str) -> Result<String, ParseError> {
+        Ok(value.to_owned())
+    }
+
+    fn parse_list(value: &str) -> Result<Vec<ListEdit<String>>, ParseError> {
+        option_value_parser::string_list_edits(value)
+            .map_err(|e| format_parse_error("string list", value, e))
+    }
+}
+
+pub(crate) trait ListMember: Parseable + DeserializeOwned {}
+
+impl ListMember for bool {}
+
+impl ListMember for i64 {}
+
+impl ListMember for f64 {}
+
+impl ListMember for String {}
+
+// If the corresponding unexpanded value points to a @fromfile, then the
+// first component is the path to that file, and the second is the value from the file,
+// or None if the file doesn't exist and the @?fromfile syntax was used.
+//
+// Otherwise, the first component is None and the second is the original value.
+type ExpandedValue = (Option<PathBuf>, Option<String>);
+
+fn mk_parse_err(err: impl Display, path: &Path) -> ParseError {
+    ParseError::new(format!(
+        "Problem reading {path} for {{name}}: {err_msg}",
+        path = path.display(),
+        err_msg = err,
+    ))
 }
 
-#[allow(dead_code)]
-pub(crate) fn parse_bool_list(value: &str) -> Result<Vec<ListEdit<bool>>, ParseError> {
-    option_value_parser::bool_list_edits(value)
-        .map_err(|e| format_parse_error("bool list", value, e))
+fn maybe_expand(value: String) -> Result<ExpandedValue, ParseError> {
```

Comment:
A helper function for the expand* functions below that are the ones actually used by args/env/config.